hivemind/client/moe.py

```diff
@@ -182,19 +184,20 @@ def forward(cls, ctx, dummy, experts_per_sample: List[List[RemoteExpert]], k_min
         mask = torch.zeros([num_samples, max_experts], dtype=torch.bool, device=flat_inputs[0].device)
         mask[alive_ii, alive_jj] = True
 
-        alive_flat_outputs_stacked = list(map(torch.cat, zip(*alive_flat_outputs)))
-        # list of torch tensors, where i-th tensor is of shape [num_responded, *expert_outputs[i].shape]
+        alive_flat_outputs_stacked = (torch.cat(outputs).to(flat_inputs[0].device)
+                                      for outputs in zip(*alive_flat_outputs))
+        # torch tensors, i-th tensor is of shape [num_responded, *expert_outputs[i].shape]
 
         outputs = []
         for response_stacked in alive_flat_outputs_stacked:
             output = torch.zeros(
-                [num_samples, max_experts, *response_stacked.shape[1:]], device=response_stacked.device,
+                [num_samples, max_experts, *response_stacked.shape[1:]], device=flat_inputs[0].device,
```

Comment:
[not critical, not sure] - it may be faster to perform __setitem__ on cpu and then send results to GPU